File: leetcode/c++/89_gray_code.cpp
// https://leetcode.cn/problems/gray-code/description

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> grayCode(int n) {
        vector<int> res;
        res.push_back(0);
        int head = 1;
        for (int i = 0; i <n; i++) {
            for (int j = res.size() - 1; j >=0; j--) 
                res.push_back(head + res[j]);
            head <<= 1;
        }
        return res;
    }
};

// n阶的格雷编码等于n-1阶的格雷编码加上n-1阶的格雷编码的逆序再在每个元素前面加1
// 时间复杂度为O(n^2)，空间复杂度为O(1)

File: leetcode/golang/89_gray_code.go
func grayCode(n int) []int {
    var res []int = make([]int, 1, 1<<n)
    res[0] = 0
    head := 1
    for i := 0; i < n; i++ {
        for j := len(res) - 1; j >= 0; j-- {
            res = append(res, head + res[j]);
        }
        head <<= 1
    }
    return res
}

// var res []int = make([]int, 1, 1<<n)，表示容量为1<<n，初始长度为1；
// var res []int = make([]int, 1<<n)，表示容量为1<<n，初始长度为1<<n